src/main/scala/sample/stream/GroupLogFile.scala

```diff
@@ -25,21 +25,25 @@ object GroupLogFile {
     // read lines from a log file
     val logFile = new File("src/main/resources/logfile.txt")
 
-    import akka.stream.io.Implicits._ // add file sources to Source object or use explicitly: SynchronousFileSource(f)
-    Source.synchronousFile(logFile).
+    FileIO.fromFile(logFile).
       // parse chunks of bytes into lines
       via(Framing.delimiter(ByteString(System.lineSeparator), maximumFrameLength = 512, allowTruncation = true)).
       map(_.utf8String).
+      map {
+        case line@LoglevelPattern(level) => (level, line)
+        case line@other => ("OTHER", line)
+      }.
       // group them by log level
-      groupBy {
-        case LoglevelPattern(level) => level
-        case other                  => "OTHER"
+      groupBy(5, _._1).
+      fold(("", List.empty[String])) {
+        case ((_, list), (level, line)) => (level, line :: list)
       }.
       // write lines of each group to a separate file
       mapAsync(parallelism = 5) {
-        case (level, groupFlow) =>
-          groupFlow.map(line => ByteString(line + "\n")).runWith(Sink.synchronousFile(new File(s"target/log-$level.txt")))
+        case (level, groupList) =>
+          Source(groupList.reverse).map(line => ByteString(line + "\n")).runWith(FileIO.toFile(new File(s"target/log-$level.txt")))
```

Comment:
Yes, this would allow the files to be written, but not the system to be shut down when the stream is finished. One problem with `toWithKey` (besides the name :-) ) is that it would be specific to groupBy, but SubFlow is reused by Split (and possibly others like `.parallel` or `.retry` that are yet to come).